saleor/discount/utils.py

```diff
@@ -102,7 +102,12 @@ def activate_voucher_code(code: "VoucherCode") -> None:
     code.save(update_fields=["is_active"])
 
 
-def add_voucher_usage_by_customer(code: "VoucherCode", customer_email: str) -> None:
+def add_voucher_usage_by_customer(
+    code: "VoucherCode", customer_email: Optional[str]
+) -> None:
+    if not customer_email:
+        raise NotApplicable("This offer is only valid for signed-in users.")
```

Comment:
It doesn't look like a truth. 👇 
![image](https://github.com/saleor/saleor/assets/23282005/e4c99a96-f8e8-4dc4-bb47-803a1c066bd4)
It looks like the offer is valid, also when you don't have an assigned order.user but you have a user_email.
So it is more like: you can't add voucher that counts a usage per customer, when you don't have user_email, or user assigned to the order. 
So the offer is also valid for non-signed-in users.